### PR TITLE
feat(SPV-846): refactor numeric error codes into strings

### PR DESCRIPTION
**Pull request by pawellewandowski98**
# Pull Request Checklist

- [ ] 📖 I created my PR using provided  : [CODE_STANDARDS](https://github.com/bitcoin-sv/spv-wallet/blob/main/.github/CODE_STANDARDS.md)
- [ ] 📖 I have read the short Code of Conduct: [CODE_OF_CONDUCT](https://github.com/bitcoin-sv/spv-wallet/blob/main/.github/CODE_OF_CONDUCT.md)
- [ ] 🏠 I tested my changes locally.
- [ ] ✅ I have provided tests for my changes.
- [ ] 📝 I have used conventional commits.
- [ ] 📗 I have updated any related documentation.
- [ ] 💾 PR was issued based on the Github or Jira issue.

# Description

- Refactored way how the errors are return. Now every error which is return from our endpoint is located in spverror package. It is required to define it in two places:
1. definitions.go - define error which will be used in code
2. errors.go - it has map which has error as a key and value contain message, status code and code error which will be returned

Example:
request does not have auth header

- Set error response by method from spverrors package
image

- Error defined in ./spverrors/errors.go 
image

- Response with status 401:
image

If error is not defined in map we return default error:
image

### CODE REVIEW

File: models/errors.go [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+type SPVError struct {
+	Code       string
+	Message    string
+	StatusCode int
+}
+
+// ResponseError is an error which will be returned in HTTP response
+type ResponseError struct {
+	Code    string `json:"Code"`
+	Message string `json:"Message"`
```
Comment (ac4ch) [inline models/errors.go:13]:
```type ResponseError struct {
	Code    string `json:"code"`    // JSON key should be in camelCase
	Message string `json:"message"` // JSON key should be in camelCase
}```
json tags

File: models/errors.go [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+}
+
+const UnknownErrorCode = "error-unknown"
```
Comment (ac4ch) [inline models/errors.go:16]: comment 
```
// UnknownErrorCode is the default error code for unknown errors
```

also not sure if just "unknown" would not be enough in the message, what u think ?

**Review by ac4ch — COMMENTED** [review]
except that one file all seems to be good i would say

File: engine/spverrors/definitions.go [hunk 3]
```diff
@@ -0,0 +1,338 @@
+package spverrors
+
+import "github.com/bitcoin-sv/spv-wallet/models"
+
+// How the Codes are generated?
+// 1. The first part of the code is the error prefix, e.g. "error"
+// 2. The second part of the code is the model name, e.g. "transaction"
+// 2. The third part of the code is the reason of the error, e.g. "not-found"
```
Comment (chris-4chain) [inline engine/spverrors/definitions.go:8]: This description doesn't align to what I see below.
Maybe sth like this:
- `"error"` - like mandatory prefix for all error codes
- (optional) `{error group}` -  e.g. `"unauthorized"`
- (optional) `{subject}` - name of model or some noun e.g. `body`, `auth-header`
- (optional) `{reason}` - what happened, e.g. `"not-found"`, `"missing"`, `"invalid"`

So the error codes would look like this:
- `error-unauthorized`
- `error-unauthorized-token-missing`
- `error-unauthorized-signature-invalid`

File: models/errors.go (hunk 2 above)
Comment (pawellewandowski98) [reply models/errors.go:16]: `error-` prefix corresponds to the structure of all error

File: engine/go.mod [hunk 4]
```diff
@@ -2,14 +2,18 @@ module github.com/bitcoin-sv/spv-wallet/engine
 
 go 1.22.4
 
+replace github.com/bitcoin-sv/spv-wallet/models => ./../models
```
[the remark this answers is not in the document]
Comment (pawellewandowski98) [reply engine/go.mod:5]: Should I add new task for that?

[the remark this answers is not in the document]
Comment (chris-4chain) [reply engine/go.mod:5]: Hmm, let's wait for Damian to discuss this. 
- I'll set a reminder to do so 👍

File: models/errors.go (hunk 1 above)
Comment (pawellewandowski98) [reply models/errors.go]: Done

File: engine/spverrors/definitions.go (hunk 3 above)
Comment (pawellewandowski98) [reply engine/spverrors/definitions.go]: Done

**Review by dorzepowski — APPROVED** [review]

**Review by chris-4chain — APPROVED** [review]

**Review by chris-4chain — APPROVED** [review]

**Pull request merged**